web/blog/2025-06-20-oauth-security-incident.md

```diff
@@ -0,0 +1,101 @@
+
+All other configurations (Google, GitHub, Discord, or Keycloak with the default case-insensitive configuration) are **not affected**. Email and username auth providers are also **not affected**.
+
+Users should upgrade to Wasp `0.16.6` which contains the fix ASAP.
+
+## Description
+
+Wasp has a concept of a `ProviderId` which contains the provider name and the provider-specific ID. 
+
+For example:
+
+- (`email`, `alice@wasp.sh`) for Email auth
+- (`google`, `10769150350006150715113082367`) for Google auth
+
+Wasp takes the provider-specific ID (the second part), converts it to a string and lowercases it to keep the IDs normalized.
+
+This approach made sense for:
+
+- **email** - emails are case insensitive (e.g. you can't signup for Gmail with `johnnY@gmail.com` if `johnny@gmail.com` already exists)
+- **usernames** - we don't want users to impersonate each other with similar usernames
+
+However, the [OpenID spec](https://openid.net/specs/openid-connect-core-1_0.html#IDToken) states the following for the `sub` property (which we use for OAuth as user ID):
+
+> REQUIRED.
+	    Subject Identifier. A locally unique and never
+	    reassigned identifier within the Issuer for the End-User,
+            …
+	    The sub value is a case-sensitive string.
+> 
+
+Treating the ID we receive as case insensitive violates the OpenID spec, but in practice, for providers that use numerical IDs, lowercasing the ID had no impact.
+
+It worked fine for:
+
+- **Google** - uses numeric ID (`10769150350006150715113082367`)
+- **GitHub** - uses numeric ID (`1`)
+- **Discord** - uses numeric ID (`80351110224678912`)
+
+It was problematic for:
+
+- **Keycloak** - uses lowercase UUID string (`25a37fd0-d10e-40ca-af6c-821f20e01be8`) by default, but you can configure Keycloak so that the IDs are case sensitive, which current Wasp auth ignores
+
+  For example, two different users with Keycloak IDs `abc` and `ABC` would be considered the same person:
+    - Example 1: `abc` person adds their credit card → `ABC` gets access to their credit card
+    - Example 2: `admin` person exists → `ADMIN` gets admin rights as well
+
+The [source](https://github.com/wasp-lang/wasp/blob/014f661a27f829bddf2290f7cdf1cd7c38f3387c/waspc/data/Generator/templates/sdk/wasp/auth/utils.ts#L85) of the problem is the `createProviderId` function that normalized the received ID:
+
+```tsx
+export function createProviderId(
+  providerName: ProviderName,
+  providerUserId: string,
+): ProviderId {
+  return {
+    providerName,
+    providerUserId: providerUserId.toLowerCase(), // <--- incorrect behavior
+  }
+}
+```
+
+## Applied fix
+
+Wasp has released version `0.16.6` which includes a fix for the vulnerability - Wasp no longer lowercases user IDs received from OAuth providers, only the `email` and `username` user IDs.
+
+All users should upgrade to Wasp `0.16.6`, especially if they use Keycloak auth with a configuration that makes the user IDs case sensitive. 
```

Comment:
```suggestion
All users should upgrade to Wasp `0.16.6`, especially if they use Keycloak auth with a configuration that makes the user IDs case sensitive. 
```
All users should upgrade, especially .. -> makes it sound like everybody it affected to some degree, but they are not, maybe rephrase somehow a bit?